File: csharp/run-length-encoding/RunLengthEncoding.cs
using System.Linq;

public static class RunLengthEncoding
{
    public static string Encode(string input)
    {
        var result = "";
        var count = 1;

        for (var i = 0; i < input.Length; i++)
        {
            if (i + 1 < input.Length && input[i] == input[i + 1])
            {
                count++;
                continue;
            }

            result += count > 1 ? count.ToString() : "";
            result += input[i];
            count = 1;
        }
        
        return result;
    }

    public static string Decode(string input)
    {
        var result = "";
        
        while (input.Any())
        {
            var digits = input.TakeWhile(char.IsDigit).ToArray();
            var count = digits.Any() ? int.Parse(string.Concat(digits)) : 1;
            var token = input.Skip(digits.Length).First();

            result += string.Concat(Enumerable.Repeat(token, count));
            input = input.Substring(digits.Length + 1);
        }

        return result;
    }
}

File: kotlin/run-length-encoding/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String = buildString {
        var remaining = input
        while (remaining.isNotEmpty()) {
            val batch = remaining.takeWhile { it == remaining.first() }

            if (batch.length > 1) {
                append(batch.length)
            }

            append(batch.first())
            remaining = remaining.drop(batch.length)
        }
    }

    fun decode(input: String): String = buildString {
        var remaining = input
        while (remaining.isNotEmpty()) {
            val digits = remaining.takeWhile { it.isDigit() }
            val char = remaining.drop(digits.length).first()
            val count = when (digits) {
                "" -> 1
                else -> digits.toInt()
            }
            append("$char".repeat(count))
            remaining = remaining.drop(digits.length + 1)
        }
    }
}
